Add variables to set region of photon emission

szhang17phys opened the pull request:
User can invoke the new variables directly at the fhicl file. Then region of photon emission can be determined based on the interest of the user, which is more flexible than the initial module, where photon emission vertices can only be spread in the whole space.

szhang17phys committed:
Delete larsim/EventGenerator/PhotonGen_module.cc

szhang17phys committed:
Add variables to set photon emission boundary

szhang17phys committed:
Delete larsim/EventGenerator/PhotonGen_module.cc

szhang17phys committed:
Add variables to set region of photon emission

User can set values of these new variables directly, in which way the region of photon emissions can be set, based on the interest of the user.

FNALbuild commented:
A new Pull Request was created by  @szhang17phys (Shuaixiang) for develop.

It involves the following packages:

larsim


@LArSoft/level-2-managers, @LArSoft/level-1-managers can you please review it and eventually sign? Thanks.

cms-bot commands are listed <a href="https://cdcvs.fnal.gov/redmine/projects/larsoft/wiki/Pull_request_comments_that_trigger_CI_actions">here</a>

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks (skipped because base branch is not develop)

lgarren commented:
code-checks

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
-code-checks
Pull request failed code-formatting checks. Please ensure that `cetmodules` has been setup and execute the following command from the top-level directory of your repository:

```
format-code \
  larsim/EventGenerator/PhotonGen_module.cc
```
Then commit the changes and push them to your PR branch.

lgarren commented:
@szhang17phys please run clang format as noted

szhang17phys committed:
Merge branch 'LArSoft:develop' into develop

FNALbuild commented:
Pull request #143 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
-code-checks
Pull request failed code-formatting checks. Please ensure that `cetmodules` has been setup and execute the following command from the top-level directory of your repository:

```
format-code \
  larsim/EventGenerator/PhotonGen_module.cc
```
Then commit the changes and push them to your PR branch.

szhang17phys committed:
clang format ran

FNALbuild commented:
Pull request #143 was updated. @LArSoft/level-2-managers, @LArSoft/level-1-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks

knoepfel commented on larsim/EventGenerator/PhotonGen_module.cc:
```diff
@@ -110,6 +118,13 @@ namespace evgen {
+    , fBmaxY{pset.get<double>("BmaxY")}
+    , fBminZ{pset.get<double>("BminZ")}
+    , fBmaxZ{pset.get<double>("BmaxZ")}
+
```
Two things:

1. I'd rename the configuration parameters (e.g.) `Xmin` instead of `BminX`.
2. Also, this change would require all users to specify `"Bmin*"` and `"Bmax*"` parameters in their configurations, and that is not always desired.  It would be better to make it an option--i.e. if the parameter `Ymin` is provided, then it should override the value of `fYmin`.

knoepfel commented on larsim/EventGenerator/PhotonGen_module.cc:
```diff
@@ -152,9 +169,23 @@ namespace evgen {
+    fYmin = fBminY;
+    fYmax = fBmaxY;
+    fZmin = fBminZ;
+    fZmax = fBmaxZ;
```
See above comment.  These assignments should only be made when the configuration parameter is provided.  Otherwise, the values of `fXmin`, `fXmax`, etc. should be set to `CryoBounds.MinX()`, etc.

knoepfel reviewed (commented):
@szhang17phys, thank you for the PR.  Please see the comments below.

szhang17phys commented:
Thanks for above nice comments! I will modify accordingly!

szhang17phys reviewed (commented):
Names of new variables modified.

szhang17phys committed:
Delete larsim/EventGenerator/PhotonGen_module.cc

FNALbuild commented:
Pull request #143 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

szhang17phys committed:
New bool-type variable UserD added

The function of UserD is to determine whether applying user-defined region or not. The default value of UserD is false. Also, default values of all user-defined boundary variables, ex: BminX, BmaxX, are all set as 0; Only when UserD is set as true, can users  set the region of their own interests. When UserD is not mentioned in fhicl script, the default region is just the boundary of whole cryostat.

FNALbuild commented:
Pull request #143 was updated. @LArSoft/level-2-managers, @LArSoft/level-1-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
-code-checks
Pull request failed code-formatting checks. Please ensure that `cetmodules` has been setup and execute the following command from the top-level directory of your repository:

```
format-code \
  larsim/EventGenerator/PhotonGen_module.cc
```
Then commit the changes and push them to your PR branch.

lgarren commented:
@szhang17phys please run the code formatting https://github.com/LArSoft/larsim/pull/143#issuecomment-2582480425

knoepfel reviewed (commented):
Thanks for the updates @szhang17phys --we're getting closer.  I was thinking of something like this:

```c++
PhotonGen::PhotonGen(fhicl::ParameterSet const& pset)
  : ...
{
  if (auto boundaries = pset.get_if_present<fhicl::ParameterSet>("boundaries")) {
    fXmin = boundaries->get<double>("Xmin");
    fXmax = boundaries->get<double>("Xmax");
    fYmin = boundaries->get<double>("Ymin");
    fYmax = boundaries->get<double>("Ymax");
    fZmin = boundaries->get<double>("Zmin");
    fZmax = boundaries->get<double>("Zmax");
  }
  else {
    art::ServiceHandle<geo::Geometry const> geo;
    auto const CryoBounds = geo->Cryostat().Boundaries();
    fXmin = CryoBounds.MinX();
    fXmax = CryoBounds.MaxX();
    fYmin = CryoBounds.MinY();
    fYmax = CryoBounds.MaxY();
    fZmin = CryoBounds.MinZ();
    fZmax = CryoBounds.MaxZ();
  }
}
```

where `boundaries` would be a nested table in the configuration:

```jsonnet
photonGen: {
  module_type: PhotonGen
  boundaries: {
    Xmin: ...
  }
}
```

If the `boundaries` parameter is in the configuration, then the module will take the parameters from the configuration; otherwise they will be provided by the geometry system.  This involves the following changes:

1. Removing the `fUserD`, `fBmin*`, and `fBmax*` variables, and
2. Moving the setting of the `fXmin` (etc.) variables to the constructor.

Let me know if you have any questions.

szhang17phys commented:
Hello! Kyle. I think I got your idea: You do not want to introduce too many variables. fXmin* will replace my fBminX* in PhotonGen::PhotonGen(fhicl::ParameterSet const& pset). If fXmin, fXmax, ... are not pointed to certain values by user, they will get values from cryoBounds automatically. I agree that your proposal is more neat! I will modify again recently!

szhang17phys committed:
Merge branch 'LArSoft:develop' into develop

FNALbuild commented:
Pull request #143 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
-code-checks
Pull request failed code-formatting checks. Please ensure that `cetmodules` has been setup and execute the following command from the top-level directory of your repository:

```
format-code \
  larsim/EventGenerator/PhotonGen_module.cc
```
Then commit the changes and push them to your PR branch.
